File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/balancedtapas/domain/BalancedTapasGame.java
package com.zwstudio.logicpuzzlesandroid.puzzles.balancedtapas.domain;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fj.F2;

public class BalancedTapasGame extends CellsGame<BalancedTapasGame, BalancedTapasGameMove, BalancedTapasGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(-1, 1),
            new Position(0, 1),
            new Position(1, 1),
            new Position(1, 0),
            new Position(1, -1),
            new Position(0, -1),
            new Position(-1, -1),
    };
    public static Position offset2[] = {
            new Position(0, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(1, 1),
    };

    public Map<Position, List<Integer>> pos2hint = new HashMap<>();
    public int left, right;

    public BalancedTapasGame(List<String> layout, String leftPart, GameInterface<BalancedTapasGame, BalancedTapasGameMove, BalancedTapasGameState> gi, GameDocumentInterface gdi) {
        super(gi, gdi);
        size = new Position(layout.size(), layout.get(0).length() / 4);
        left = right = leftPart.charAt(0) - '0';
        if (leftPart.length() > 1) left++;
        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                String s = str.substring(c * 4, c * 4 + 4).trim();
                if (s.isEmpty()) continue;
                List<Integer> hint = new ArrayList<>();
                for (char ch : s.toCharArray()) {
                    if (ch == '?' || ch >= '0' && ch <= '9') {
                        int n = ch == '?' ? -1 : (ch - '0');
                        hint.add(n);
                    }
                }
                pos2hint.put(p, hint);
            }
        }
        BalancedTapasGameState state = new BalancedTapasGameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(BalancedTapasGameMove move, F2<BalancedTapasGameState, BalancedTapasGameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        BalancedTapasGameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean switchObject(BalancedTapasGameMove move) {
        return changeObject(move, (state, move2) -> state.switchObject(move2));
    }

    public boolean setObject(BalancedTapasGameMove move) {
        return changeObject(move, (state, move2) -> state.setObject(move2));
    }

    public BalancedTapasObject getObject(Position p) {
        return state().get(p);
    }

    public BalancedTapasObject getObject(int row, int col) {
        return state().get(row, col);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/balancedtapas/domain/BalancedTapasGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.balancedtapas.domain

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class BalancedTapasGame(layout: List<String>, leftPart: String, gi: GameInterface<BalancedTapasGame, BalancedTapasGameMove, BalancedTapasGameState>, gdi: GameDocumentInterface) : CellsGame<BalancedTapasGame, BalancedTapasGameMove, BalancedTapasGameState>(gi, gdi) {
    companion object {
        val offset = arrayOf(
            Position(-1, 0),
            Position(-1, 1),
            Position(0, 1),
            Position(1, 1),
            Position(1, 0),
            Position(1, -1),
            Position(0, -1),
            Position(-1, -1)
        )
        val offset2 = arrayOf(
            Position(0, 0),
            Position(0, 1),
            Position(1, 0),
            Position(1, 1)
        )
    }

    var pos2hint = mutableMapOf<Position, List<Int>>()
    var left = 0
    var right = 0

    init {
        size = Position(layout.size, layout[0].length / 4)
        left = leftPart[0] - '0'
        right = left
        if (leftPart.length > 1) left++
        for (r in 0 until rows()) {
            val str = layout[r]
            for (c in 0 until cols()) {
                val p = Position(r, c)
                val s = str.substring(c * 4, c * 4 + 4).trim(' ')
                if (s.isEmpty()) continue
                val hint = mutableListOf<Int>()
                for (ch in s)
                    when (ch) {
                        in '0'..'9' -> hint.add(ch - '0')
                        '?' -> hint.add(-1)
                    }
                pos2hint[p] = hint
            }
        }
        val state = BalancedTapasGameState(this)
        states.add(state)
        levelInitilized(state)
    }

    private fun changeObject(move: BalancedTapasGameMove, f: (BalancedTapasGameState, BalancedTapasGameMove) -> Boolean): Boolean {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size).clear()
            moves.subList(stateIndex, states.size).clear()
        }
        val state = cloner.deepClone(state())
        val changed = f(state, move)
        if (changed) {
            states.add(state)
            stateIndex++
            moves.add(move)
            moveAdded(move)
            levelUpdated(states[stateIndex - 1], state)
        }
        return changed
    }

    fun switchObject(move: BalancedTapasGameMove) = changeObject(move, BalancedTapasGameState::switchObject)
    fun setObject(move: BalancedTapasGameMove) = changeObject(move, BalancedTapasGameState::setObject)
    fun getObject(p: Position) = state()[p]
    fun getObject(row: Int, col: Int) = state()[row, col]
}
